cmd/oras/internal/option/target.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package option
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/spf13/pflag"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content/oci"
+	"oras.land/oras-go/v2/registry"
+)
+
+const (
+	TargetTypeRemote    = "remote"
+	TargetTypeOCILayout = "oci"
+)
+
+// Unary target option struct.
+type Target struct {
+	Remote
+	RawReference string
+	Type         string
+	TagOrDigest  string
+
+	isOCI bool
+}
+
+// ApplyFlags applies flags to a command flag set for unary target
+func (opts *Target) ApplyFlags(fs *pflag.FlagSet) {
+	opts.applyFlagsWithPrefix(fs, "", "")
+	opts.Remote.ApplyFlags(fs)
+}
+
+// AnnotatedReference returns full printable reference.
+func (opts *Target) AnnotatedReference() string {
+	return fmt.Sprintf("[%s] %s", opts.Type, opts.RawReference)
+}
+
+// applyFlagsWithPrefix applies flags to a command flag set with a prefix string.
+// Commonly used for non-unary remote targets.
+func (opts *Target) applyFlagsWithPrefix(fs *pflag.FlagSet, prefix, description string) {
+	var (
+		flagPrefix string
+		noteSuffix string
+	)
+	if prefix != "" {
+		flagPrefix = prefix + "-"
+		noteSuffix = description + " "
+	}
+	fs.BoolVarP(&opts.isOCI, flagPrefix+"oci", "", false, "Set "+noteSuffix+"target as an OCI-layout.")
```

Comment:
I really like using uri formats for the source and destination like `oras cp file://my-directory oci://localhost/owner/name:v1`

Reply:
Also using a bool here kind of gives a proliferation of arguments that would happen if there was just a `--source=format` and `--destination=format` option. In my world format would be either file or registry.

Reply:
The Matt's over at helm tend to be pretty pendantic, I doubt they made up the `oci://` format although I cannot find an RFC.

Reply:
URI schemes are a bit hard to be integrated with oras in terms of design.

Firstly, we don't have a standard scheme for OCI Layout (in a tar or as a folder), OCI-compliant Registry, Docker-compliant Registry, Docker Daemon, and so on in the [IANA registry](https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml). It means that the same URI scheme can refer to different concept in various tools. Some may interpret `oci` as OCI Layout in a tar, some may state that `oci` means OCI-compliant registry. Again, there is no standard.

Secondly, the UX should be consistent. If we introduce something like
```sh
oras cp file://my-directory oci://localhost/owner/name:v1
```
we have to unify all the UX to be something like
```sh
oras cp docker://docker.io/library/alpine:latest oci://localhost:5000/alpine:latest

# or using script
src=docker.io/library/alpine:latest
dst=localhost:5000/alpine:latest
oras cp docker://$src oci://$dst
```
assuming `docker` means Docker-compliant registry and `oci` means OCI-compliant registry. The UX is much more complex than
```sh
oras cp docker.io/library/alpine:latest localhost:5000/alpine:latest

# or using script
src=docker.io/library/alpine:latest
dst=localhost:5000/alpine:latest
oras cp $src $dst
```

If we really want the URI scheme, we could do something like
```sh
oras cp --from-target type=oci,ref=uri file://./my-image-directory:latest localhost:5000/owner/name:v1

# or other protocols
oras cp --from-target type=oci,ref=uri ftp://localhost/my-repo-dir:latest localhost:5000/owner/name:v1
oras cp --from-target type=oci,ref=uri https://localhost/my-repo-dir:latest localhost:5000/owner/name:v1
```

Reply:
okay, thanks for the thoughtful responses. I get now that the format of the output cannot really be encapsulated in a URI. To me, that doesn't completely rule out using `file://`, but from what you are both saying `oci://` doesn't make a lot of sense. I like the idea that the default source and target would be an oci registry and any registry would be supported in that format `localhost:5000/owner/name:v1`